File: HashTable/LC138-CopyListWithRandomPointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head: return None
        
        d = dict()
        p = head
        
        while p:
            new_node = Node(p.val, None, None)
            d[p] = new_node
            p = p.next
            
        p = head
        
        while p:
            # p is the original Node
            # d[p] is the new Node
            # d[p.next] is the next Node
            if p.next:
                d[p].next = d[p.next]
            if p.random:
                d[p].random = d[p.random]
            p = p.next
        
        return d[head]
            

File: LinkedList/LC138-CopyListWithRandomPointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (!head) return head;
        
        unordered_map<Node*, Node*> mp;
        Node* cur = mp[head] = new Node(head->val);
        Node* res = cur;
        while (head) {
            if (head->random) {
                auto it = mp.find(head->random);
                if (it == mp.end()) {
                    it = mp.emplace(head->random, new Node(head->random->val)).first;
                }
                cur->random = it->second;
            }
            if (head->next) {
                auto it = mp.find(head->next);
                if (it == mp.end()) {
                    it = mp.emplace(head->next, new Node(head->next->val)).first;
                }
                cur->next = it->second;
            }
            cur = cur->next;
            head = head->next;
        }
        return res;
    }
};
